### PR TITLE
[#1345] Authorship: Add last modified date to each LoC if specified

### LINKED ISSUE #1345: Authorship.json: give per-line last-modified-date
Current:
image

I'm wondering is it possible to also give last-modified-date for each line. While not strictly not necessary for the dashboard, I can use that data in further processing when I grade students' work.

This will increase the size of the file significantly though. Perhaps we can disable by default with an option to enable it using a flag.

The time zone (if provided) needs to be considered as well.

### PR DESCRIPTION
**Pull request by jamessspanggg**
Fixes #1345 

```
The last modified date of each line of code is not present in the
authorship.json.

It would be beneficial for tutors to use the last modified date of
each line of code to do further processing.

Let's add the last modified date of each line of code within the
authorship.json. This functionality will only be enabled when the
user specifies the flag '--last-modified-date' or '-l' in short.
```

Snapshot of updated `authorship.json`:
Screen Shot 2020-10-27 at 8 31 01 PM

### CODE REVIEW

File: src/systemtest/java/reposense/ConfigSystemTest.java
Comment (jamessspanggg) [inline src/systemtest/java/reposense/ConfigSystemTest.java:88]: Added another system test for this functionality, however, I feel that this is redundant and takes up space. Should I merge this test with the `testDateRange` as one?

**Comment by jamessspanggg** [comment]
Hi Prof @damithc, 
- would it be okay if i use the `--last-modified-date` or `-l` to toggle this feature? the feature will be disabled by default.

With using the following args `--config ./config -v -s d1`,
- performance wise: 33.06 second(s) (without the `-l`), 35.06 second(s) (with the `-l`)
- total size of authorship.json: 190kb (without the `-l`), 213kb (with the `-l`)

File: src/systemtest/java/reposense/ConfigSystemTest.java
Comment (jamessspanggg) [inline src/systemtest/java/reposense/ConfigSystemTest.java:78]: Created this system test that merges with the previous `testDateRange` as it feels redundant to have another system test that may take up our build time and space.

**Comment by Tejas2805** [comment]
> * performance wise: 33.06 second(s) (without the `-l`), 35.06 second(s) (with the `-l`)
> * total size of authorship.json: 190kb (without the `-l`), 213kb (with the `-l`)

Which repo are the above parameters for?

**Comment by jamessspanggg** [comment]
> Which repo are the above parameters for?
> 
> 
> 
> 

The ones in our repo config.

**Comment by damithc** [comment]
> Hi Prof @damithc,
> 
> * would it be okay if i use the `--last-modified-date` or `-l` to toggle this feature? the feature will be disabled by default.

Sounds good. In the documentation, mention `lower case L` to prevent someone reading it as upper case i.

**Comment by jamessspanggg** [comment]
> Sounds good. In the documentation, mention `lower case L` to prevent someone reading it as upper case i.

Noted prof. Have done so.

File: docs/ug/cli.md
Comment (Tejas2805) [inline docs/ug/cli.md:22]: Just realised. We say same command as above but don't include `--timezone`. Can include it?

File: src/main/java/reposense/authorship/FileInfoAnalyzer.java
Comment (Tejas2805) [inline src/main/java/reposense/authorship/FileInfoAnalyzer.java:85]: Not sure if this change is required. Since it's technically a part of the `aggregateBlame`. @fzdy1914 your thoughts?

**Review by Tejas2805 — COMMENTED** [review]
Left 2 comments. 

Just one quick thought: Should we include in our documentation that it will affect the performance and size of the json file? Though it's obvious since extra stuff is being added, but might be a good thing to mention?

Everything else LGTM!

File: docs/ug/cli.md
Comment (jamessspanggg) [reply docs/ug/cli.md:22]: I think that can be included in #1346 as it is more relevant to this change

File: src/main/java/reposense/authorship/FileInfoAnalyzer.java
Comment (jamessspanggg) [reply src/main/java/reposense/authorship/FileInfoAnalyzer.java:85]: how about `aggregateBlameInfo`?

File: docs/ug/cli.md
Comment (dcshzj) [reply docs/ug/cli.md:22]: Yep, I think the last modified date needs to be made timezone-aware as well

Comment (jamessspanggg) [reply docs/ug/cli.md:22]: Yep, placing this PR on hold until #1346 merges

Comment (dcshzj) [reply docs/ug/cli.md:22]: #1346 was merged, I think we can move ahead with this PR?

**Comment by jamessspanggg** [comment]
@dcshzj I refactored some parts of ur code to `TimeUtil` as there are multiple parts utilising the same method.

Since you have worked with timezone related issues, could you check if my implementation related to timezones are ok? especially the `convertDateWithZones` method.

File: src/main/java/reposense/util/TimeUtil.java
Comment (dcshzj) [inline src/main/java/reposense/util/TimeUtil.java:75]: I think the variable should not use `systemRawOffset` as it can be confused as referring to the actual system time zone.

File: src/main/java/reposense/util/TimeUtil.java
Comment (dcshzj) [inline src/main/java/reposense/util/TimeUtil.java:73]: We might be able to use the functionality in this function for `getSystemDateFromZonedDate`, since it is using the same logic.

File: src/main/java/reposense/authorship/FileInfoAnalyzer.java
Comment (dcshzj) [inline src/main/java/reposense/authorship/FileInfoAnalyzer.java:109]: I can't remember exactly, but if `commitDateInMs` is a number that is the number of milliseconds from epoch, then we need to convert from the system time zone to the user-specified time zone, not from the author's time zone.

The milliseconds represents a point in time, which is converted to a `Date` object based on the system time zone. We then convert the date (in the system time zone) to the user-specified time zone so that when we drop the hour, minute and seconds, the year, month and day is still correct.

**Review by dcshzj — COMMENTED** [review]
> @dcshzj I refactored some parts of ur code to `TimeUtil` as there are multiple parts utilising the same method.
> 
> Since you have worked with timezone related issues, could you check if my implementation related to timezones are ok? especially the `convertDateWithZones` method.

For `ArgsParserTest.java`, there is also a `getZoneRawOffset` method but it is in the test suite. Should we also extract it out, since the test you introduced in `FileAnalyzerTest.java` will also require adapting it to the correct timezone?

File: src/main/java/reposense/util/TimeUtil.java
Comment (dcshzj) [reply src/main/java/reposense/util/TimeUtil.java:64]: I think we should stick to using `getSystemDateFromZonedDate`? Right now I think the `fromZoneId` will always be the system default, which was the original implementation.

File: src/main/java/reposense/util/TimeUtil.java
Comment (dcshzj) [inline src/main/java/reposense/util/TimeUtil.java:64]: I think we should stick to using `getSystemDateFromZonedDate`? Right now the `fromZoneId` will always be the system default, which was the original implementation. I don't think we will have a case where the `Date` object is not in the system time zone.

**Review by dcshzj — APPROVED** [review]
LGTM 👍

File: src/main/java/reposense/authorship/FileInfoAnalyzer.java
Comment (fzdy1914) [inline src/main/java/reposense/authorship/FileInfoAnalyzer.java:108]: Should the method be `getZonedDateFromSystemDate`?

File: src/main/java/reposense/model/LocationsCliArguments.java
Comment (fzdy1914) [inline src/main/java/reposense/model/LocationsCliArguments.java:16]: Change the variable name to `isLastModifiedDateIncluded` to match the other variable name?

File: src/main/java/reposense/authorship/FileInfoAnalyzer.java
Comment (jamessspanggg) [reply src/main/java/reposense/authorship/FileInfoAnalyzer.java:108]: @dcshzj may i get ur previous rationale in naming the method?

Comment (dcshzj) [reply src/main/java/reposense/authorship/FileInfoAnalyzer.java:108]: Yes, I think it might be more appropriate to rename it to `getZonedDateFromSystemDate`. The Javadoc comment will need to be updated accordingly.

File: src/main/java/reposense/authorship/FileInfoAnalyzer.java
Comment (fzdy1914) [inline src/main/java/reposense/authorship/FileInfoAnalyzer.java:107]: from system default time zone to cli specified timezone?

File: src/main/java/reposense/commits/CommitResultAggregator.java
Comment (fzdy1914) [inline src/main/java/reposense/commits/CommitResultAggregator.java:156]: Can double confirm the usage of this method?

**Review by fzdy1914 — COMMENTED** [review]
Some comment

File: src/main/java/reposense/commits/CommitResultAggregator.java
Comment (fzdy1914) [reply src/main/java/reposense/commits/CommitResultAggregator.java:156]: My bad for not looking into the detail of these two methods in the previous PR.

Comment (dcshzj) [reply src/main/java/reposense/commits/CommitResultAggregator.java:156]: `getSystemStartOfDate` gets the instant where it says 00:00:00 in the system time zone for the given date. `getStartOfDate` gets the instant where it says 00:00:00 in the given `zoneId` time zone for the given date.

We still need the system time zone because Gson does not allow for a time zone to be specified, so they only recognise the system default time zone.

**Pull request merged**
